sigsched.3
.TH sigsched 3
.SH NAME
sigsched \- signal-schedule (non-preemptive threads) library
.SH SYNTAX
.B #include <sigsched.h>

ss_sig *\fBss_asap()\fR;
.br
ss_sig *\fBss_signal(\fIsigno\fB)\fR;
.br
ss_sig *\fBss_sigread(\fIfd\fB)\fR;
.br
ss_sig *\fBss_sigwrite(\fIfd\fB)\fR;
.br
ss_sig *\fBss_sigexcept(\fIfd\fB)\fR;

int \fBss_addsig(\fIsigno\fB)\fR;

int \fBss_schedvwait(\fIsig,t,flagi,i,p,wait\fB)\fR;
.br
int \fBss_schedwait(\fIsig,t,i,wait\fB)\fR;
.br
int \fBss_sched(\fIsig,t,i\fB)\fR;

int \fBss_schedonce(\fIsig,t,i\fB)\fR;

int \fBss_unschedv(\fIsig,t,flagi,i,p\fB)\fR;
.br
int \fBss_unsched(\fIsig,t,i\fB)\fR;

void \fBss_externsetsig(\fIsig,x\fB)\fR;

int \fBss_exec()\fR;

void \fBss_forcewait()\fR;
.br
void \fBss_unforcewait()\fR;

ss_thread \fI*t\fP;
.br
ss_sig \fI*sig\fP;
.br
ss_extern \fI*x\fP;
.br
int \fIflagi\fP;
.br
int \fIsigno\fP;
.br
int \fIfd\fP;
.br
ss_id \fIi\fP;
.br
ss_idptr \fIp\fP;
.br
int \fIwait\fP;
.SH DESCRIPTION
.B sigsched
implements the signal-schedule programming model,
otherwise known as non-preemptive threads,
otherwise known as event-based programming.
A thread is scheduled to execute upon receipt of a signal
(occurrence of an event).
Separate threads do not interrupt each other.
All they can do is schedule more threads.

.B sigsched
supports far more flexible signals than C normally provides
under UNIX.
``File descriptor 2 is writable'' is a signal, for example.
Furthermore, threads do not have to be written to handle a
signal at any moment, so code written to use
.B sigsched
can be fully optimized.
In contrast, preemptive thread models (including
UNIX's usual signal handling) prevent optimizations involving global
variables.

In general, a ``signal'' is any persistent condition.
The ``file descriptor 2 is writable'' signal starts when the pipe
is created, persists at least until the next I/O, finishes when the pipe is
written to capacity, restarts when the pipe is read, and so on.
UNIX signals are examples of
.I thread-lowered signals.
For example, SIGINT starts (is raised) when some process executes
kill(pid,SIGINT),
and finishes (is lowered) just before process pid calls the appropriate
signal handler (thread).
Note that if another process calls
kill(pid,SIGINT)
before the first one is delivered,
the signal merely persists.
It is not delivered twice, as after the first delivery the
signal condition has been turned off and can't be redelivered.
Any number of kill()s may be absorbed into
one delivery in this way.

With
.B sigsched,
the program can schedule a thread to execute upon receipt of a signal.
.B ss_schedvwait()
and
.B ss_unschedv()
schedule and unschedule threads.
.B ss_exec()
then executes one scheduled thread after another, as described below.
It exits when there are no ``important'' threads left to execute.

.B ss_schedvwait(\fIsig,t,flagi,i,p,wait\fB)
schedules the thread
.I t
to execute with integer identifier
.I i
or pointer identifier
.I p
as soon as condition
.I sig
exists.
This is an ``important'' thread if
.I wait
is nonzero.
.I sig
is of type
.B ss_sig *;
various functions produce signals of this type.
.I t
is of type
.B ss_thread *,
defined as a function returning void;
it is called as
.I t(i)
if
.I flagi
is nonzero,
or
.I t(p)
if
.I flagi
is zero.
.I i
is an integer,
which must be zero if
.I flagi
is;
.I p
is a character pointer,
which must be a null pointer if
.I flagi
is nonzero.
.B <sigsched.h>
defines the
.B ss_sig
and
.B ss_thread
types;
it also abbreviates
int as
.B ss_id
and char * as
.B ss_idptr.
.B ss_schedvwait
normally returns 0, but will return -1
in case of a memory allocation failure.

.B ss_unschedv(\fIsig,t,flagi,i,p\fB)
unschedules the thread
.I t
previously scheduled to execute with identifier
.I i
or
.I p
as soon as condition
.I sig
existed.
.I flagi,
.I i,
and
.I p
must follow the same rules as above.
.B ss_unschedv
returns 0 if the unschedule was successful,
1 if there was no such thread.
The effects are currently undefined if a thread is scheduled
more than once for the same signal with the same identifier.

.B ss_exec()
executes one thread after another, with no interruptions.
It calls
.I t(id)
only if, for some signal
.I sig,
(1)
.I t(id)
is scheduled to execute upon
.I sig;
(2) condition
.I sig
has existed sometime between the end of the last call
of a thread scheduled upon
.I sig
and the beginning of this call to
.I t(id).
If a
thread has just finished executing and
.B ss_exec
can call one or more
threads under the above restrictions, it will choose one and call that,
unless every scheduled thread has a wait value of 0
(i.e., there are no important threads scheduled).
In the latter case
(including, for example, when there are no threads scheduled at all),
.B ss_exec()
immediately returns 0.
It returns -1 in case of a memory allocation
failure; in that case its internal structures may be permanently
corrupted, and
.B ss_exec
may not be called again in the same program.

If no threads can execute at a given moment,
but if some thread is
scheduled with a non-zero wait value,
.B ss_exec
has to wait for a signal
to arrive.
As an optimization,
it will block the process until some
thread can execute,
rather than actively polling the set of signals.

Note that if several threads are scheduled to execute upon one signal,
and the signal suddenly exists, one of the threads will execute.
If the
signal turns off before the end of that thread, the other threads
scheduled upon the signal will not execute.
This is always true for
thread-lowered signals.
This behavior stands in
marked contrast to the behavior of interrupts---upon an interrupt,
all the scheduled threads would be executed.

Each signal provides its own scheduling guarantees.
For instance, under
this implementation,
any (waiting) thread scheduled on the signal
.I ss_asap()
will in fact execute at some point, provided that no thread
blocks or loops forever.
There is no way to keep pushing 
.I ss_asap()
farther and farther into the future by scheduling other threads.
On the other hand,
.I ss_asap()
will never flush out the other builtin signals.

.B sigsched
provides several builtin signals:
.B ss_asap()
returns a (pointer to a) signal which always exists.
.B ss_signal(\fIsigno\fB)
returns a thread-lowered signal which is true when UNIX signal
.I signo
is received.
.B ss_sigread(\fIfd\fB)
returns a signal which is true when
.I fd
is open and readable, as defined by
.I select();
similarly for
.B ss_sigwrite
and
.B ss_sigexcept.

In order for
.B sigsched
to handle UNIX signal
.I signo,
you must call
.B ss_addsig(\fIsigno\fB)
before calling
.B ss_exec().
.B ss_addsig
will discard the old signal handler;
later,
.B ss_exec
will not restore the handler upon exiting, and may
leave the signal blocked or unblocked.
.B ss_addsig
will return 0 normally,
-1 if
.I signo
is not in the right range for signals.
If another library makes use of
.B ss_signal
with
.B sigsched,
it should provide a mandatory initialization routine
which calls
.B ss_addsig.

.B ss_schedvwait
and
.B ss_unschedv
can be abbreviated in common cases.
.B ss_schedwait(\fIsig,t,i,wait\fB)
is the same as
.B ss_schedvwait(\fIsig,t,1,i,(ss_idptr)0,wait\fB).
.B ss_sched(\fIsig,t,i\fB)
is the same with
.I wait
set to 0; it is commonly used for
handling user signals.
.B ss_unsched(\fIsig,t,i\fB)
is the same as
.B ss_unschedv(\fIsig,t,1,i,(ss_idptr)0\fB).

.B ss_schedonce(\fIsig,t,i\fB)
is similar to
.B ss_sched
but is in fact implemented on top of
.B ss_schedvwait
with an independent mechanism.
Each call to
.B ss_schedonce
schedules
.I t
upon a new signal which starts when
.I sig
does and exists only until
.I t(i)
is executed.
After the first execution the new signal disappears.
The new signal cannot be unscheduled.

.B ss_forcewait()
tells
.B sigsched
that something important is going on outside
.B sigsched
and that
.B ss_exec
should not exit.
.B ss_unforcewait()
negates a previous
.B ss_forcewait().
.B ss_forcewait()
and
.B ss_unforcewait()
control a counter, not a flag, so independent
libraries can use them, but each library should
be careful to use as many of one call as of the other.
These functions must not be used outside
.B ss_exec().

.B ss_externsetsig(sig,x)
creates a new signal
in the
.B ss_sig
pointed to by
.I sig.
.I x
points to an
.B ss_extern,
which is defined as follows in
.B <sigsched.h>:
.PP
.EX
typedef struct {
  int (*sched)();
  int (*unsched)();
  union { int n; char *c; } u;
} ss_extern;
.EE
.PP
.I sched
must be filled in with a scheduling function,
which is called as
.I (*sched)(x,t,flagi,i,p,wait)
whenever
.B ss_schedvwait(\fIsig,t,flagi,i,p,wait\fB)
is called;
similarly for
.I unsched.
Use of
.I u
is up to the caller.
.I sched
and
.I unsched
must observe the same rules as
.B ss_schedvwait
and
.B ss_unschedv
on any other signals: i.e., they must schedule threads upon
a persistent condition, make sure that
.I ss_exec
does not exit if
any important threads are scheduled, etc.
Note that
.B ss_externsetsig
records
.I x
in
.I sig,
so
.I x
must point either to static memory or to
memory which remains allocated as long as
any thread is scheduled or executing upon
.I sig.
Memory management of the
.I sig
structure itself is up to the caller.

It is recommended that library
.I foo
define a
.B foo_sig
structure, which contains
.B ss_sig
.I sig,
.B ss_extern
.I x,
and any other necessary information for the signals defined by
.I foo.
Then
.B foo_setsig(\fI&fsig,otherinfo\fB),
where
.I fsig
is a
.B foo_sig,
should set up the
.I otherinfo,
set
.I fsig.x.u.c
to
.I &fsig,
set
.I fsig.x.sched
and
.I fsig.x.unsched
appropriately,
and
finish with
.B ss_externsetsig(&fsig.sig,&fsig.x).
That way the user can use
.I &fsig.sig
as the signal argument to
.B sigsched
functions,
and when
.I foo's
scheduling routines are passed
.I &fsig.x
as a first argument,
they can get to
.I otherinfo
through
.I fsig.x.u.c.

.B sigsched
uses
.B ralloc
for all allocation.
.SH VERSION
sigsched 1.1, August 25, 1991.
.SH AUTHOR
Placed into the public domain by Daniel J. Bernstein.
.SH "SEE ALSO"
select(2),
sigvec(2),
ralloc(3)


CHANGES
2/9/92: pty 4.0.
2/9/92: Fixed up documentation, hopefully for the last time.
1/29/92: Changed INSTALL.c to report chown root for INSTALL root. Tnx IS.
1/27/92: Added ABRT->IOT ifndef in sigdfl.c.
1/27/92: Changed sigdfl.c to zero the sigvecs rather than setting sv_flags.
1/16/92: pty 4.0, second gamma.
1/13/92: Does it work on an Iris? Nobody knows. 
1/13/92: Added IRIS_UGH_PTYS to ptyget.c. Tnx IRIS, I think.
1/13/92: Changed fatal to warning for tctpgrp failure in ptyslave.c. Tnx IRIS.
1/13/92: ifndef'ed SYSV for ut_host in ptylogs.c. Tnx IRIS.
1/13/92: Added config/sysv.h to distribution. Tnx IRIS.
1/13/92: Added PTYLIBS, noted its use in README. Tnx IRIS.
1/13/92: Replaced CCOPTMZ with OPTMZ.
1/13/92: Noted in CALLS that TTY_TERMIO seems to work. Tnx IRIS.
1/13/92: Adapted tty_initmodes for TTY_TERMIO. Ugh, ugh. Tnx IRIS.
1/13/92: Adapted tty_mungemodes for TTY_TERMIO. Ugh. Tnx IRIS.
1/13/92: Adapted tty_modifymodes for TTY_TERMIO (poorly). Tnx IRIS.
1/13/92: Adapted tty_getmodes and tty_setmodes for TTY_TERMIO. Tnx IRIS.
1/13/92: In ptytty.c used ifdefs for FIONREAD and TIOCEXCL. Tnx IRIS.
1/13/92: Adapted ptytty.h for TTY_TERMIO. Tnx IRIS.
1/13/92: Added TTY_TERMIO option to config/ttyopts.h. Tnx IRIS.
1/13/92: Gave in and made tty_copymodes do a struct copy.
1/13/92: Fixed typo in SECURITY. Tnx KS.
1/13/92: Fixed typo in pty-basic.1. Tnx KS.
1/12/92: Replaced <sgtty.h> with <sys/ioctl.h> in ptytty.h. Tnx IRIS.
1/12/92: Added chmod 700 for all shell scripts in Makefile.
1/12/92: Removed <sys/timeb.h> from wall.c and write.c. Tnx IRIS.
1/12/92: In separate effort got it working on an IRIS. See above.
12/22/91: pty 4.0, first gamma! (Still have to integrate BSD 4.4 version.)
12/22/91: Changed various things in the documentation.
12/22/91: Fixed exit code reporting for, e.g., DIE_NOPTYS.
12/22/91: Changed POSIX to POSIX_SILLINESS. See config/posix.h.
12/20/91: Added CCOPTMZ in Makefile for radixsort.
12/19/91: Added BLURB.
12/19/91: Added TESTS.
12/19/91: Cleaned up README.
12/19/91: Added WTF, noted SunOS 4.0.3 forcefg bug in README.
12/18/91: Fixed bug in sessmenu's sesslist parsing.
12/17/91: Fixed up documentation a bit; referred to checkptys in README.
12/16/91: Added CPP define to Makefile.
12/15/91: Added -xi.
12/15/91: Started machine list in Makefile.
12/15/91: Added EWOULDBLOCK in ptysecure.c for Domain/OS.
12/14/91: Added target-list target.
12/14/91: Added lint target.
12/13/91: Changed master so that it continues child on any reconnect. Tnx SS.
11/24/91: Changed master to run only as effective uid.
11/18/91: Added TERM, USR1, USR2 ignoring to master. Tnx SS.
11/9/91: Added \fP in syntax sections in pty-opts.1, pty.1. Tnx SS.
11/9/91: Added <sys/types.h> before <utmp.h> in {users,who}.c. Domain/OS idiocy.
11/9/91: Fixed dependencies for CHECKCONF and INSTALL.
11/3/91: Added proper .h dependencies in Makefile (finally).
10/27/91: Added exit(0) at end of sesslist.c. Tnx SS.
10/27/91: Fixed ptysecure.c---it didn't undo chmod for some setups. Tnx SS.
10/27/91: Added make clean. Tnx SS.
10/27/91: Added checkptys.8.
10/27/91: Changed checkptys to be installed non-setuid.
10/26/91: Added pty.1, pty-basic.1, pty-opts.1, sess.1, condom.1.
10/26/91: Tossed old man pages.
10/26/91: Added DIE_EXIST for ptysigler.c.
10/26/91: Added DIE_ELSE for ptysigler.c.
10/26/91: Fixed ptymain to die(SETMODES) in an impossible situation.
10/26/91: Changed a few die() values in ptymain.c.
10/25/91: Need to add *.h to Makefile...
10/25/91: Removed last traces of old getopt schizophrenia.
10/25/91: Changed *rotate to preserve modes of log files.
10/25/91: Added sessnowinit, scnowinit.
10/25/91: Added sessrotate, sclogrotate.
10/25/91: Fixed off-by-one on remotelen.
10/25/91: Changed wtmprotate to depend on config/wtmpfile.h properly.
[log is not complete before this point]
10/6/91: Added -xsS.
9/30/91: pty 4.0, second beta.


sigsched.c
/* sigsched.c, sigsched.h: signal-schedule thread library
Daniel J. Bernstein, brnstnd@nyu.edu.
Depends on ralloc.h, sod.h, config/fdsettrouble.h.
Requires BSDish environment: reliable signals, sig{vec,block,setmask}, select.
9/1/91: Added worst-case fdset, FD_ZERO, etc. definitions.
8/25/91: sigsched 1.1, public domain.
8/25/91: Fixed bug that sigs[sched->blah].r didn't force instant timeout.
8/25/91: Fixed bug that if select() returned -1 then fds were still checked.
7/21/91: Changed forever to a 1-hour wakeup.
7/19/91: Added isopen() to fix bug in case of bad descriptor.
7/18/91: Baseline. sigsched 1.0, public domain.
No known patent problems.

Documentation in sigsched.3.

XXX: how well do we clean up upon ss_exec() exit?

*/

#include <sys/types.h>
#include <sys/param.h>
#include <sys/time.h>
#include <signal.h>
#include <errno.h>
#include <fcntl.h>
extern int errno;
#include "config/fdsettrouble.h"
#include "sigsched.h"
#include "ralloc.h"
#include "sod.h"

/* XXX: should restore signal set exactly after ss_exec returns */

typedef int sigc_set; /*XXX */

#define sigc_ismember(x,i) (*(x) & (1 << ((i) - 1)))
#define sigc_addset(x,i) (*(x) |= (1 << ((i) - 1)))
#define sigc_emptyset(x) (*(x) = 0)

/*       sigprocmask(SIG_UNBLOCK,xxxx,(sigc_set *) 0); */
#define sigc_unblock(x) (sigsetmask(sigblock(0) & ~*(x)))
/*       sigprocmask(SIG_BLOCK,xxxx,(sigc_set *) 0); */
#define sigc_block(x) (sigblock(*(x)))

#ifndef NSIG
#define NSIG 64 /* it's not as if any sane system has more than 32 */
#endif

#define NUMSIGS NSIG

#ifndef FD_SETSIZE
#define FD_SETSIZE 256
#endif

#define NUMFDS FD_SETSIZE /* if select() can't handle it, we can't either */

#ifdef LACKING_FD_ZERO
#define NFDBITS	(sizeof(fd_mask) * NBBY)
#define	FD_SET(n,p) ((p)->fds_bits[(n)/NFDBITS] |= (1 << ((n) % NFDBITS)))
#define	FD_ISSET(n,p) ((p)->fds_bits[(n)/NFDBITS] & (1 << ((n) % NFDBITS)))
#define FD_ZERO(p) bzero((caddr_t)(p),sizeof(*(p)))
#endif

#ifdef DESPERATION_FD_SET
#undef NFDBITS
#undef FD_SET
#undef FD_ISSET
#undef FD_ZERO
#undef fd_set
#define fd_set long
#define FD_SET(n,p) ((*p) |= (1 << (n)))
#define FD_ISSET(n,p) ((*p) & (1 << (n)))
#define FD_ZERO(p) (*p = 0L)
#endif

#define ASAP 1
#define SIGNAL 2
#define READ 3
#define WRITE 4
#define EXCEPT 5
#define JUNK 6
#define EXTERN 7

typedef struct { ss_sig s; int r; } ss_sigplus;

static ss_sigplus asap;
static ss_sigplus sigs[NUMSIGS];
static ss_sigplus reads[NUMFDS];
static ss_sigplus writes[NUMFDS];
static ss_sigplus excepts[NUMFDS];
static ss_sigplus junk; /* special case for internal use */

static void initsigs()
{
 int i;
 asap.s.type = ASAP; asap.s.u.n = 0; asap.r = 0;
 for (i = 0;i < NUMSIGS;++i)
  { sigs[i].s.type = SIGNAL; sigs[i].s.u.n = i; sigs[i].r = 0; }
 for (i = 0;i < NUMFDS;++i)
  { reads[i].s.type = READ; reads[i].s.u.n = i; reads[i].r = 0; }
 for (i = 0;i < NUMFDS;++i)
  { writes[i].s.type = WRITE; writes[i].s.u.n = i; writes[i].r = 0; }
 for (i = 0;i < NUMFDS;++i)
  { excepts[i].s.type = EXCEPT; excepts[i].s.u.n = i; excepts[i].r = 0; }
 junk.s.type = JUNK; junk.s.u.n = 0; junk.r = 0;
}

ss_sig *ss_asap()
{ return &(asap.s); }
#define OKsig(i) ((i >= 0) && (i < NUMSIGS))
ss_sig *ss_signal(i) int i;
{ if (!OKsig(i)) return 0; return &(sigs[i].s); }
#define OKfd(fd) ((fd >= 0) && (fd < NUMFDS))
ss_sig *ss_sigread(fd) int fd;
{ if (!OKfd(fd)) return 0; return &(reads[fd].s); }
ss_sig *ss_sigwrite(fd) int fd;
{ if (!OKfd(fd)) return 0; return &(writes[fd].s); }
ss_sig *ss_sigexcept(fd) int fd;
{ if (!OKfd(fd)) return 0; return &(excepts[fd].s); }

void ss_externsetsig(sig,x)
ss_sig *sig;
ss_extern *x;
{
 sig->type = EXTERN;
 sig->u.c = (char *) x;
}

struct sched
 {
  ss_sig *sig;
  ss_thread *t;
  union { ss_id i; ss_idptr p; } id;
  int flagi;
  int wait;
 }
;

SODdecl(schedlist,struct sched);

static schedlist schedhead = 0;
static int schednum = 0;
static int schedjunked = 0;
static int numwait = 0;

void ss_forcewait()
{
 ++numwait;
}

void ss_unforcewait()
{
 --numwait;
}

int ss_schedvwait(sig,t,flagi,i,p,wait)
ss_sig *sig;
ss_thread *t;
int flagi;
ss_id i;
ss_idptr p;
int wait;
{
 schedlist s;

 if (sig->type == EXTERN)
  {
   ss_extern *x;
   x = (ss_extern *) sig->u.c;
   return x->sched(x,t,flagi,i,p,wait);
  }
 s = SODalloc(schedlist,s,ralloc);
 if (!s)
   return -1;
 SODdata(s).sig = sig;
 SODdata(s).t = t;
 if (SODdata(s).flagi = flagi)
   SODdata(s).id.i = i;
 else
   SODdata(s).id.p = p;
 SODdata(s).wait = wait;
 SODpush(schedhead,s);
 ++schednum;
 if (wait)
   ++numwait;
 return 0;
}

int ss_schedwait(sig,t,i,wait)
ss_sig *sig;
ss_thread *t;
ss_id i;
int wait;
{
 return ss_schedvwait(sig,t,1,i,(ss_idptr) 0,wait);
}

int ss_sched(sig,t,i)
ss_sig *sig;
ss_thread *t;
ss_id i;
{
 return ss_schedvwait(sig,t,1,i,(ss_idptr) 0,0);
}

struct oncestuff { ss_sig *sig; ss_thread *t; ss_id i; } ;
/* XXX: this is the same as some other struct */

static void once(p)
ss_idptr p;
{
 struct oncestuff *os;
 os = (struct oncestuff *) p;
 if (ss_unschedv(os->sig,once,0,0,p) == -1)
   ; /* impossible */
 os->t(os->i);
 RFREE(os);
}

int ss_schedonce(sig,t,i)
ss_sig *sig;
ss_thread *t;
ss_id i;
{
 struct oncestuff *os;

 os = (struct oncestuff *) ralloc(sizeof(struct oncestuff));
 if (!os)
   return -1;
 os->sig = sig; os->t = t; os->i = i;
 return ss_schedvwait(sig,once,0,0,(ss_idptr) os,1);
}

/* XXX: could rallocinstall() this, if it has the recvhead() test */

static int schedcleanup()
{
 schedlist s;
 schedlist t;
 schedlist sprev;

/*  if (recvhead) return 0;  XXX: needs recvhead in scope */

 if (!schedjunked)
   return 0;

 sprev = 0;
 s = schedhead;
 while (s)
  {
   if (SODdata(s).sig == &(junk.s))
    {
     if (sprev)
      {
       SODpop(SODnext(sprev),t); /* XXX: not part of official sod interface */
       s = SODnext(sprev);
      }
     else
      {
       SODpop(s,t);
       schedhead = s;
      }
     SODfree(t,rfree);
     --schednum;
     --schedjunked;
    }
   else
    {
     sprev = s;
     s = SODnext(s);
    }
  }

/* schednum -= schedjunked; now done dynamically inside loop */
/* schedjunked = 0; */
 return 1;
}

static void nothing(id)
ss_id id;
{
 ;
}

int ss_unschedv(sig,t,flagi,i,p)
ss_sig *sig;
ss_thread *t;
int flagi;
ss_id i;
ss_idptr p;
{
 schedlist s;

 if (sig->type == EXTERN)
  {
   ss_extern *x;
   x = (ss_extern *) sig->u.c;
   return x->unsched(x,t,flagi,i,p);
  }
 for (s = schedhead;s;s = SODnext(s))
   if (SODdata(s).sig == sig && SODdata(s).t == t)
     if (SODdata(s).flagi == flagi)
       if (flagi ? (SODdata(s).id.i == i) : (SODdata(s).id.p == p))
        {
         SODdata(s).sig = &(junk.s);
         SODdata(s).t = nothing; /* just in case */
         if (SODdata(s).wait)
           --numwait;
         SODdata(s).wait = 0;
         ++schedjunked;
         return 0;
        }
 return 1;
}

int ss_unsched(sig,t,i)
ss_sig *sig;
ss_thread *t;
ss_id i;
{
 return ss_unschedv(sig,t,1,i,(ss_idptr) 0);
}

static struct timeval timeout;
static struct timeval instant = { 0, 0 };
static struct timeval forever = { 3600, 0 };
  /* XXX: talk to me */

static void handle(i)
int i;
{
 timeout = instant; /* XXX: structure copying */
 sigs[i].r = 1;
}

static sigc_set sigstorage;
static sigc_set *xxxx = 0;

int ss_addsig(i)
int i;
{
 if (!OKsig(i))
   return -1;
 if (!xxxx)
  {
   xxxx = &sigstorage;
   sigc_emptyset(xxxx);
  }
 sigc_addset(xxxx,i);
 return 0;
}

static int isopen(fd)
int fd;
{
 /* XXX: should call this only if select() fails */
 return fcntl(fd,F_GETFL,0) != -1;
}

SODdecl(recvlist,schedlist);

int ss_exec()
{
 int i;
 struct sigvec sv;
 recvlist recvhead;
 recvlist temp;
 schedlist sch;

 initsigs();

 if (xxxx)
  {
   sigc_block(xxxx);

   sv.sv_handler = handle;
   sv.sv_mask = *xxxx; /* so handle won't interrupt itself */
   sv.sv_flags = 0;

   /* XXX: Does anyone but me find it absolutely idiotic that POSIX
      doesn't provide a way to get each member of a signal set in turn? */
   for (i = 0;i < NUMSIGS;i++)
    {
     if (sigc_ismember(xxxx,i))
       if (sigvec(i,&sv,(struct sigvec *) 0) == -1) /*XXX: really trash orig? */
	 ; /* not our problem */
    }
  }

 recvhead = 0;

 while (numwait)
  {
   if (recvhead)
    {
     int w;
     SODpop(recvhead,temp);
     sch = SODdata(temp);

/* This is the only section where we call user code. */
#define DOIT \
if (SODdata(sch).flagi) \
  SODdata(sch).t(SODdata(sch).id.i); \
else \
  SODdata(sch).t(SODdata(sch).id.p);

     switch(SODdata(sch).sig->type)
      {
       case JUNK:
	 break; /* has been unscheduled while waiting on the receive list */
       case ASAP:
	 DOIT
	 break;
       case READ:
	 if (reads[w = SODdata(sch).sig->u.n].r)
	   DOIT
	 reads[w].r = 0;
	 break;
       case WRITE:
	 if (writes[w = SODdata(sch).sig->u.n].r)
	   DOIT
	 writes[w].r = 0;
	 break;
       case EXCEPT:
	 if (excepts[w = SODdata(sch).sig->u.n].r)
	   DOIT
	 excepts[w].r = 0;
	 break;
       case SIGNAL:
	 if (sigs[w = SODdata(sch).sig->u.n].r)
	   DOIT
	 sigs[w].r = 0;
	   /* ``after the end of the last...'' */
	 break;
       case EXTERN:
	 /* by definition, an external library handles this */
	 break;
       default: /* XXX: huh? */
	 ;
      }
     SODfree(temp,rfree);
    }
   else
    {
     schedlist sp;
     static fd_set rfds;
     static fd_set wfds;
     static fd_set efds;
     static int maxfd;
     int r;

     if (schedjunked > 100)
       if (schednum / schedjunked < 3)
         (void) schedcleanup(); /* now's as good a time as any */

     timeout = forever;
     FD_ZERO(&rfds);
     FD_ZERO(&wfds);
     FD_ZERO(&efds);
     maxfd = -1;

     for (sp = schedhead;sp;sp = SODnext(sp))
      {
       switch(SODdata(sp).sig->type)
	{
	 case JUNK:
	   break;
	 case ASAP:
	   timeout = instant;
	   break;
	 case SIGNAL:
	   if (sigs[SODdata(sp).sig->u.n].r)
	     timeout = instant;
	   break;
	 case READ:
	   if (isopen(SODdata(sp).sig->u.n))
	     FD_SET(SODdata(sp).sig->u.n,&rfds);
	   if (SODdata(sp).sig->u.n > maxfd)
	     maxfd = SODdata(sp).sig->u.n;
	   break;
	 case WRITE:
	   if (isopen(SODdata(sp).sig->u.n))
	     FD_SET(SODdata(sp).sig->u.n,&wfds);
	   if (SODdata(sp).sig->u.n > maxfd)
	     maxfd = SODdata(sp).sig->u.n;
	   break;
	 case EXCEPT:
	   if (isopen(SODdata(sp).sig->u.n))
	     FD_SET(SODdata(sp).sig->u.n,&efds);
	   if (SODdata(sp).sig->u.n > maxfd)
	     maxfd = SODdata(sp).sig->u.n;
	   break;
	 case EXTERN:
	   break;
	 default: /*XXX: huh? */
	   break;
	}
      }

     if (xxxx)
       sigc_unblock(xxxx);
     /* This is the only section where handle() can be called. */
     /* XXX: If maxfd == -1, this select functions as a pause. */
     /* XXX: If maxfd == -1 and timeout is instant, should skip select. */
     /* XXX: Random bug of note: Real BSD systems will say that the
        fd is writable as soon as a network connect() fails. The first
	I/O will show the error (though it's rather stupid that you
	can't find out the error without doing I/O). What does Ultrix
	4.1 do? It pauses for 75 seconds. Dolts. */
     r = select(maxfd + 1,&rfds,&wfds,&efds,&timeout);
       /* XXX: does this necessarily prevent timeout race conditions? */
     if (xxxx)
       sigc_block(xxxx);

     if (r == -1)
      {
       FD_ZERO(&rfds);
       FD_ZERO(&wfds);
       FD_ZERO(&efds);
       switch(errno)
        {
         case EINTR: /* fine, this will happen on any signal */ break;
         case EBADF: /* who knows? */ break;
         case EINVAL: /* simply impossible */ break;
         default: /*XXX*/ ;
	 /* well, that was real useful */
        }
      }

     for (sp = schedhead;sp;sp = SODnext(sp))
      {
       switch(SODdata(sp).sig->type) 
	{
	 case JUNK:
	   break;
	 case ASAP:
           temp = SODalloc(recvlist,temp,ralloc);
           if (!temp)
             return -1; /*XXX*/
	   SODdata(temp) = sp;
           SODpush(recvhead,temp);
	   break;
	 case SIGNAL:
	   if (sigs[SODdata(sp).sig->u.n].r)
	    {
             temp = SODalloc(recvlist,temp,ralloc);
             if (!temp)
               return -1; /*XXX*/
	     SODdata(temp) = sp;
             SODpush(recvhead,temp);
	    }
	   break;
	 case READ:
	   if (FD_ISSET(SODdata(sp).sig->u.n,&rfds))
	    {
	     FD_CLR(SODdata(sp).sig->u.n,&rfds);
	     reads[SODdata(sp).sig->u.n].r = 1;
             temp = SODalloc(recvlist,temp,ralloc);
             if (!temp)
               return -1; /*XXX*/
	     SODdata(temp) = sp;
             SODpush(recvhead,temp);
	    }
	   break;
	 case WRITE:
	   if (FD_ISSET(SODdata(sp).sig->u.n,&wfds))
	    {
	     FD_CLR(SODdata(sp).sig->u.n,&wfds);
	     writes[SODdata(sp).sig->u.n].r = 1;
             temp = SODalloc(recvlist,temp,ralloc);
             if (!temp)
               return -1; /*XXX*/
	     SODdata(temp) = sp;
             SODpush(recvhead,temp);
	    }
	   break;
	 case EXCEPT:
	   if (FD_ISSET(SODdata(sp).sig->u.n,&efds))
	    {
	     FD_CLR(SODdata(sp).sig->u.n,&efds);
	     excepts[SODdata(sp).sig->u.n].r = 1;
             temp = SODalloc(recvlist,temp,ralloc);
             if (!temp)
               return -1; /*XXX*/
	     SODdata(temp) = sp;
             SODpush(recvhead,temp);
	    }
	   break;
	 case EXTERN:
	   break;
	 default:
	   break;
	}
      }
    }
  }
 if (xxxx)
   sigc_unblock(xxxx);
   /* XXX: should put this at other returns as well */
 return 0;
}
